File: services/agent-service/app/routers/agent.py
from fastapi import APIRouter, HTTPException

from app.config import get_settings
from app.schemas import (
    ChatRequest,
    ChatResponse,
    GenerateArtifactRequest,
    GenerateArtifactResponse,
    ProvidersResponse,
)
from app.services.artifact_generator import ArtifactGeneratorService
from app.services.chat_service import ChatArtifactService
from app.services.providers.base import ArtifactGenerationError
from app.services.providers.factory import list_providers
from app.services.tool_client import ToolServiceError

router = APIRouter(prefix="/agent", tags=["agent"])


@router.get("/providers", response_model=ProvidersResponse)
def get_providers() -> ProvidersResponse:
    return list_providers(get_settings())


@router.post("/generate-artifact", response_model=GenerateArtifactResponse)
def generate_artifact(request: GenerateArtifactRequest) -> GenerateArtifactResponse:
    generator = ArtifactGeneratorService(get_settings())
    try:
        return generator.generate(request)
    except ArtifactGenerationError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    except ToolServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.post("/chat", response_model=ChatResponse)
def chat(request: ChatRequest) -> ChatResponse:
    chat_service = ChatArtifactService(get_settings())
    try:
        return chat_service.chat(request)
    except ArtifactGenerationError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    except ToolServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc

File: services/agent-service/src/routes/agent.ts
import type { FastifyInstance } from 'fastify';
import type { AppConfig } from '../config.js';
import { parseChatRequest, parseGenerateArtifactRequest, ValidationError } from '../schemas.js';
import { ArtifactGeneratorService } from '../services/artifact-generator.js';
import { ChatArtifactService } from '../services/chat-service.js';
import { ArtifactGenerationError, listProviders } from '../services/providers.js';

export function registerAgentRoutes(fastify: FastifyInstance, config: AppConfig): void {
  fastify.get('/agent/providers', async () => listProviders(config));

  fastify.post('/agent/generate-artifact', async (request, reply) => {
    let parsed;
    try {
      parsed = parseGenerateArtifactRequest(request.body);
    } catch (err) {
      if (err instanceof ValidationError) {
        return reply.code(422).send({ detail: err.message });
      }
      throw err;
    }

    try {
      return await new ArtifactGeneratorService(config).generate(parsed);
    } catch (err) {
      if (err instanceof ArtifactGenerationError) {
        return reply.code(502).send({ detail: err.message });
      }
      throw err;
    }
  });

  fastify.post('/agent/chat', async (request, reply) => {
    let parsed;
    try {
      parsed = parseChatRequest(request.body);
    } catch (err) {
      if (err instanceof ValidationError) {
        return reply.code(422).send({ detail: err.message });
      }
      throw err;
    }

    try {
      return await new ChatArtifactService(config).chat(parsed);
    } catch (err) {
      if (err instanceof ArtifactGenerationError) {
        return reply.code(502).send({ detail: err.message });
      }
      throw err;
    }
  });
}
